## Libraries imports

In [ ]:
pip install tabula-py

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.0/12.0 MB 62.3 MB/s eta 0:00:00


In [ ]:
import pandas as pd
import numpy as np
import tabula

## Data imports

In [ ]:
data = pd.read_csv("/content/drive/MyDrive/Statapp/first_drop_data.csv")

<ipython-input-3-dbf0f8ef4c9e>:1: DtypeWarning: Columns (10823) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv("/content/drive/MyDrive/Statapp/first_drop_data.csv")


In [ ]:
data.head()

HHIDPN  S1HHIDPN  R1MSTAT  R1MPART  S1BMONTH  S1BYEAR  S1BDATE  S1BFLAG  \
0      1010       0.0      5.0      0.0       NaN      NaN      NaN      NaN   
1      2010       0.0      7.0      0.0       NaN      NaN      NaN      NaN   
2      3010    3020.0      1.0      0.0       9.0   1938.0  -7778.0      0.0   
3      3020    3010.0      1.0      0.0       1.0   1936.0  -8752.0      0.0   
4  10001010       0.0      8.0      0.0       NaN      NaN      NaN      NaN   

   S1COHBYR  S1HRSAMP  ...  genetic_4_PP_COGENT17  genetic_4_SBP_COGENT17  \
0       NaN       NaN  ...                    NaN                     NaN   
1       NaN       NaN  ...                    NaN                     NaN   
2       3.0       1.0  ...               -0.58662                -0.66936   
3       3.0       1.0  ...               -0.06204                -0.70205   
4       NaN       NaN  ...                    NaN                     NaN   

   genetic_4_EGFR_CKDGEN19  genetic_4_EGFRTE_CKDGEN19  \
0                      NaN                        NaN   
1                      NaN                        NaN   
2                  1.04423                    1.02390   
3                 -1.06050                   -1.08282   
4                      NaN                        NaN   

   genetic_4_EA3_W23_SSGAC18  genetic_4_HBA1CAA_MAGIC17  \
0                        NaN                        NaN   
1                        NaN                        NaN   
2                    1.53372                    0.47688   
3                    0.85439                   -0.62488   
4                        NaN                        NaN   

   genetic_4_HBA1CEA_MAGIC17  genetic_4_GCOG2_CHARGE18  genetic_VERSION  \
0                        NaN                       NaN              NaN   
1                        NaN                       NaN              NaN   
2                   -1.28083                  -1.51568              4.3   
3                   -1.29031                   0.40118              4.3   
4                        NaN                       NaN              NaN   

   genetic_Section_A_or_E  
0                     NaN  
1                     NaN  
2                       E  
3                       E  
4                     NaN  

[5 rows x 10824 columns]

In [ ]:
# Only if necessary. (It appears to ALWAYS be necessary :-( )

# convert PDF into CSV
tabula.convert_into("/content/drive/MyDrive/Statapp/Documentation_données_socio_économique.pdf", "Documentation_données_socio_économique.csv", output_format="csv", pages='all')

Mar 26, 2023 11:00:22 PM org.apache.pdfbox.pdmodel.font.FileSystemFontProvider <init>
Mar 26, 2023 11:00:22 PM org.apache.pdfbox.pdmodel.font.FileSystemFontProvider <init>
Mar 26, 2023 11:00:22 PM org.apache.pdfbox.pdmodel.font.PDType1Font <init>
Mar 26, 2023 11:00:22 PM org.apache.pdfbox.pdmodel.font.PDType1Font <init>
Mar 26, 2023 11:04:17 PM org.apache.pdfbox.pdmodel.font.PDSimpleFont toUnicode



In [ ]:
columns_types = pd.read_csv("Documentation_données_socio_économique.csv", skiprows=299)

<ipython-input-6-90cce48ecf6b>:1: DtypeWarning: Columns (6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42) have mixed types. Specify dtype option on import or set low_memory=False.
  columns_types = pd.read_csv("Documentation_données_socio_économique.csv", skiprows=299)


In [ ]:
columns_types.sample(5)

Household Identifier . . . .         .  \
28727                                           H1AFRLES     12652   
42798                                                 10  R10HERET   
48941                                           S11JLIFT      5269   
31023  ______________________________________________...       NaN   
16894  .V=Spouse NR | 200 417 536 311 220 380 317 365...       NaN   

                                                 ..1    ..2  ..3  ..4  ..5  \
28727                                           4.90   2.05  1.0  9.0  NaN   
42798  R10HERET:W10 R covered in retirement /summary  Categ  NaN  NaN  NaN   
48941                                           3.32   0.98  1.0  5.0  NaN   
31023                                            NaN    NaN  NaN  NaN  NaN   
16894                                            NaN    NaN  NaN  NaN  NaN   

       ..6  ..7  ..8  ... ..31 ..32 ..33 ..34 ..35 ..36 ..37 ..38 ..39 . 60  
28727  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
42798  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
48941  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
31023  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
16894  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[5 rows x 43 columns]

## Getting types of columns

In [ ]:
columns_types = columns_types.rename(columns={"Household Identifier . . . .": "Wave", ".":"Name", "..2":"Type1", "..3": "Type2"})

columns_types = columns_types[["Name", "Type1", "Type2"]]
columns_types["Type1"] = columns_types["Type1"].fillna(columns_types["Type2"])
columns_types = columns_types.drop(columns=["Type2"]).rename(columns={"Type1": "Type"})

columns_types = columns_types.loc[columns_types["Type"].isin(["Categ", "Cont", "Char"])]

columns_types = columns_types.reset_index(drop=True)

In [ ]:
columns_types.head()

Name  Type
0    HHIDPN  Cont
1      HHID  Char
2        PN  Char
3  RAHHIDPN  Char
4    H1HHID  Cont

We now assert that each column types has been recorded. 

In [ ]:
columns_not_recorded = []
for col in data.columns:
  if "genetic" not in col and col not in columns_types["Name"].values:
    columns_not_recorded.append(col)

print(f"Among the {data.shape[1]} columns in our data, {len(columns_not_recorded)} are of unknown type.")

Among the 10824 columns in our data, 456 are of unknown type.


Still 456 columns are to handle. Let us see if we could say something about their type, in view of the number of different values they take.

In [ ]:
uniques_number = {}
for col in columns_not_recorded:
  x = data[col].nunique()
  if x not in uniques_number.keys():
    uniques_number[x] = 1
  else:
    uniques_number[x] += 1
uniques_number = dict(sorted(uniques_number.items()))
for k, v in uniques_number.items():
  print(f"{v} column(s) have exactly {k} values." )

51 column(s) have exactly 2 values.
24 column(s) have exactly 3 values.
48 column(s) have exactly 4 values.
48 column(s) have exactly 5 values.
126 column(s) have exactly 6 values.
30 column(s) have exactly 7 values.
3 column(s) have exactly 8 values.
1 column(s) have exactly 9 values.
2 column(s) have exactly 11 values.
1 column(s) have exactly 13 values.
1 column(s) have exactly 14 values.
1 column(s) have exactly 17 values.
1 column(s) have exactly 18 values.
1 column(s) have exactly 19 values.
1 column(s) have exactly 20 values.
1 column(s) have exactly 21 values.
1 column(s) have exactly 23 values.
1 column(s) have exactly 36 values.
1 column(s) have exactly 39 values.
1 column(s) have exactly 63 values.
1 column(s) have exactly 64 values.
1 column(s) have exactly 69 values.
1 column(s) have exactly 70 values.
1 column(s) have exactly 74 values.
1 column(s) have exactly 78 values.
1 column(s) have exactly 79 values.
1 column(s) have exactly 80 values.
1 column(s) have exactly 81 v

By checking the documentation, we see that columns which present more thant 17 unique values are continuous and the ones with less than 7 values are categorical. Let save this.

The doubt remains for columns which contain between 8 and 16 unique values. Let us explore them by hand.

In [ ]:
for col in columns_not_recorded:
  if data[col].nunique() <= 17 and data[col].nunique() >= 8:
    print(col, data[col].unique())

REXITWV [ 3.  6. 12. 13. nan  5. 11.  9.  2. 10.  7.  4.  8. 14.]
REXITYR [1996. 2002. 2014. 2016.   nan 2000. 2012. 2008. 1994. 2010. 2004. 1998.
 2006. 2018. 1995.]
R9DRINKN [nan  1.  0.  3.  2.  5.  4. 10.  6. 12.  7.  8. 15. 11. 16. 17.  9. 14.]
S1WORKLM [ nan  30.  50.   0. 100.  20.  60.  10.  80.  40.  70.  90.]
S13COHORT [nan  3.  4.  1.  0.  2.  5.  6.  7.]
S14COHORT [nan  3.  4.  1.  0.  2.  5.  6.  7.]
R8RETDTWV [nan  8.  7.  6.  1.  5.  3.  4.  2.]
R9RETDTWV [nan  8.  7.  6.  1.  5.  9.  4.  3.  2.]
R1WORKLM [ 50.  nan  30.   0. 100.  20.  40.  80.  60.  90.  10.  70.]


After checking documentation:
- REXITWV, R8RETDTWV, R9RETDTWV are categorical variables
- REXITYR, R9DRINKN, S1WORKLM, R1WORKLM are continuous variables
- S13COHORT and S14COHORT are somehow not found in documentation

In [ ]:
other_columns = {"Name" : ["REXITWV", "R8RETDTWV", "R9RETDTWV", "REXITYR", "R9DRINKN", "S1WORKLM", "R1WORKLM"],
                 "Type" : ["Categ", "Categ", "Categ", "Cont", "Cont", "Cont", "Cont"]}
for col in columns_not_recorded:
  number_of_values = data[col].nunique()

  if number_of_values >= 17:
    other_columns["Name"].append(col)
    other_columns["Type"].append("Cont")

  elif number_of_values <= 7:
    other_columns["Name"].append(col)
    other_columns["Type"].append("Categ")

columns_types = pd.concat([columns_types, pd.DataFrame(other_columns)]).reset_index(drop=True)

In [ ]:
print(columns_types.shape)
columns_types.head()

(14902, 2)


Name  Type
0    HHIDPN  Cont
1      HHID  Char
2        PN  Char
3  RAHHIDPN  Char
4    H1HHID  Cont

## Exporting types of columns

In [ ]:
columns_types.to_csv("columns_types.csv")

## Creating dummy variables

As many of our variables are categorical, performing direct imputation would lead to meaningless results: a variable 'example' that take [a, b, c, d, e] as values may be imputed 2.25 or any other unwanted numeric value. 

To deal with this problem, we will convert each categorical variable into a list of dummy variables. In our previous exemple, we would replace the sole variable that take [a, b, c, d, e] as values by five variables 'example_a', 'example_b', 'example_c', 'example_d', 'example_e' that takes 0 or 1 as values. In this paradigm, having an imputed value of example_e = 0.87 could be seen as the probabilty for 'example' to take value 'e'.

In [ ]:
categorical_variables = columns_types.loc[columns_types["Type"] == "Categ"]["Name"].values
remaining_categorical_variables = [col for col in categorical_variables if (col in data.columns and data[col].nunique() != 2)]

In [ ]:
len(remaining_categorical_variables)

3898

In [ ]:
data_dummies = pd.get_dummies(data, columns=remaining_categorical_variables)
print(data.shape, data_dummies.shape)
data_dummies.head()

(42233, 10824) (42233, 28791)


HHIDPN  S1HHIDPN  R1MPART  S1BMONTH  S1BYEAR  S1BDATE  S1BFLAG  S1HRSAMP  \
0      1010       0.0      0.0       NaN      NaN      NaN      NaN       NaN   
1      2010       0.0      0.0       NaN      NaN      NaN      NaN       NaN   
2      3010    3020.0      0.0       9.0   1938.0  -7778.0      0.0       1.0   
3      3020    3010.0      0.0       1.0   1936.0  -8752.0      0.0       1.0   
4  10001010       0.0      0.0       NaN      NaN      NaN      NaN       NaN   

   S1AHDSMP  S1HISPAN  ...  R13IADL5H_2  R13IADL5H_3  R13IADL5H_4  \
0       NaN       NaN  ...            0            0            0   
1       NaN       NaN  ...            0            0            0   
2       0.0       0.0  ...            0            0            0   
3       0.0       0.0  ...            0            0            0   
4       NaN       NaN  ...            0            0            0   

   R13IADL5H_5  R14IADL5H_0  R14IADL5H_1  R14IADL5H_2  R14IADL5H_3  \
0            0            1            0            0            0   
1            0            1            0            0            0   
2            0            1            0            0            0   
3            0            1            0            0            0   
4            0            1            0            0            0   

   R14IADL5H_4  R14IADL5H_5  
0            0            0  
1            0            0  
2            0            0  
3            0            0  
4            0            0  

[5 rows x 28791 columns]

In [ ]:
for col in remaining_categorical_variables:
  data_dummies.loc[data[col].isna(), data_dummies.columns.str.startswith(col)] = np.nan
data_dummies.head()

HHIDPN  S1HHIDPN  R1MPART  S1BMONTH  S1BYEAR  S1BDATE  S1BFLAG  S1HRSAMP  \
0      1010       0.0      0.0       NaN      NaN      NaN      NaN       NaN   
1      2010       0.0      0.0       NaN      NaN      NaN      NaN       NaN   
2      3010    3020.0      0.0       9.0   1938.0  -7778.0      0.0       1.0   
3      3020    3010.0      0.0       1.0   1936.0  -8752.0      0.0       1.0   
4  10001010       0.0      0.0       NaN      NaN      NaN      NaN       NaN   

   S1AHDSMP  S1HISPAN  ...  R13IADL5H_2  R13IADL5H_3  R13IADL5H_4  \
0       NaN       NaN  ...          0.0          0.0          0.0   
1       NaN       NaN  ...          0.0          0.0          0.0   
2       0.0       0.0  ...          0.0          0.0          0.0   
3       0.0       0.0  ...          0.0          0.0          0.0   
4       NaN       NaN  ...          0.0          0.0          0.0   

   R13IADL5H_5  R14IADL5H_0  R14IADL5H_1  R14IADL5H_2  R14IADL5H_3  \
0          0.0          1.0          0.0          0.0          0.0   
1          0.0          1.0          0.0          0.0          0.0   
2          0.0          1.0          0.0          0.0          0.0   
3          0.0          1.0          0.0          0.0          0.0   
4          0.0          1.0          0.0          0.0          0.0   

   R14IADL5H_4  R14IADL5H_5  
0          0.0          0.0  
1          0.0          0.0  
2          0.0          0.0  
3          0.0          0.0  
4          0.0          0.0  

[5 rows x 28791 columns]

## Export data

In [ ]:
data_dummies.to_csv("data_dummies.csv", index=False)